File: backend/src/services/counterfactual.ts
import dayjs from 'dayjs';
import type {
  AppSettings,
  BenchmarkEtf,
  CounterfactualPoint,
  CounterfactualResult,
  Instrument,
  Transaction,
} from '@decisionguru/shared';
import { countryFromSymbol } from '@decisionguru/shared';
import { getFxRate, ensureFxRange } from './fx.js';
import { getHistory, getDividends, getQuote, priceOn, ensureHistory } from './marketdata.js';
import { dividendTax, wealthTax, fundIncomeTaxDrag } from './tax.js';
import { xirr, cagr, yearsBetween, type CashFlow } from './math.js';

function resolveBenchmark(symbol: string, settings: AppSettings): BenchmarkEtf {
  const found = settings.benchmarks.find((b) => b.symbol === symbol);
  if (found) return found;
  return {
    symbol,
    name: symbol,
    currency: 'USD',
    domicile: countryFromSymbol(symbol) ?? 'US',
    incomeYield: settings.tax.defaultEtfIncomeYield,
    accumulating: false,
  };
}

interface PriceLookup {
  ccy: string;
  priceCHFOn: (date: string) => Promise<number>;
}

async function makeLookup(symbol: string, from: string, fallbackCcy: string): Promise<PriceLookup> {
  await ensureHistory(symbol, from);
  const q = await getQuote(symbol).catch(() => null);
  const ccy = q?.currency || fallbackCcy;
  return {
    ccy,
    priceCHFOn: async (date: string) => {
      const p = await priceOn(symbol, date);
      if (p == null) return 0;
      const fx = await getFxRate(ccy, 'CHF', date);
      return p * fx;
    },
  };
}

/**
 * Compare an actual holding to the counterfactual where every cash outflow into the
 * stock had instead bought the benchmark ETF on the same date. All CHF, after Swiss tax.
 * See docs/TAX-MODEL.md.
 */
export async function computeCounterfactual(
  instrument: Instrument,
  txs: Transaction[],
  benchmarkSymbol: string,
  settings: AppSettings,
  preTax = false,
  opts: { asOf?: string | null } = {},
): Promise<CounterfactualResult> {
  const tax = settings.tax;
  const bench = resolveBenchmark(benchmarkSymbol, settings);
  // 'asOf' evaluates the basket as of a chosen date instead of today (scenario date control).
  const today = opts.asOf || dayjs().format('YYYY-MM-DD');
  // Exclude corporate actions (swaps/delistings) from the counterfactual cash-flow mirror.
  const sorted = [...txs]
    .filter((t) => t.category !== 'corporate_action' && t.date <= today)
    .sort((a, b) => a.date.localeCompare(b.date));
  const buys = sorted.filter((t) => t.action === 'buy');
  const sells = sorted.filter((t) => t.action === 'sell');

  if (!buys.length) {
    return emptyResult(bench);
  }

  const firstDate = sorted[0].date;
  // Fully closed if all bought quantity has been sold.
  const boughtQty = buys.reduce((s, t) => s + t.quantity, 0);
  const soldQty = sells.reduce((s, t) => s + t.quantity, 0);
  const fullyClosed = soldQty >= boughtQty - 1e-9 && sells.length > 0;
  let endDate = fullyClosed ? sells[sells.length - 1].date : today;
  if (dayjs(endDate).isAfter(dayjs(today))) endDate = today;

  // Pre-fetch FX series for the currencies involved to keep sampling fast.
  const currencies = new Set<string>(['CHF', instrument.currency, bench.currency]);
  txs.forEach((t) => currencies.add(t.currency));
  await ensureFxRange([...currencies], firstDate, endDate);

  const stockLook = await makeLookup(instrument.symbol, firstDate, instrument.currency);
  const etfLook = await makeLookup(bench.symbol, firstDate, bench.currency);

  // --- Build ETF counterfactual units by mirroring each buy's CHF outflow ---
  interface UnitEvent {
    date: string;
    units: number;
    cashCHF: number;
  }
  const etfEvents: UnitEvent[] = [];
  const etfFlowsAfterTax: CashFlow[] = [];
  for (const b of buys) {
    const costOrig = b.quantity * b.unitPrice + (b.fees ?? 0);
    const fxBuy = await getFxRate(b.currency || instrument.currency, 'CHF', b.date);
    const cashCHF = costOrig * fxBuy;
    const etfPriceCHF = await etfLook.priceCHFOn(b.date);
    const units = etfPriceCHF > 0 ? cashCHF / etfPriceCHF : 0;
    etfEvents.push({ date: b.date, units, cashCHF });
    etfFlowsAfterTax.push({ date: b.date, amount: -cashCHF });
  }
  const totalInvestedCHF = etfEvents.reduce((s, e) => s + e.cashCHF, 0);

  const unitsHeldAt = (date: string) =>
    etfEvents.filter((e) => e.date <= date).reduce((s, e) => s + e.units, 0);
  const totalEtfUnits = unitsHeldAt(endDate);

  // ETF distributions received over the hold (distributing funds), taxed as income.
  const etfDivs = await getDividends(bench.symbol, firstDate);
  let etfNetDistCHF = 0;
  let etfGrossDistCHF = 0;
  for (const d of etfDivs) {
    if (d.date < firstDate || d.date > endDate) continue;
    const units = unitsHeldAt(d.date);
    if (units <= 0) continue;
    const fx = await getFxRate(bench.currency, 'CHF', d.date);
    const grossCHF = units * d.close * fx;
    etfGrossDistCHF += grossCHF;
    const bd = dividendTax(grossCHF, bench.domicile, tax);
    etfNetDistCHF += preTax ? grossCHF : bd.netAfterTaxCHF;
    etfFlowsAfterTax.push({ date: d.date, amount: preTax ? grossCHF : bd.netAfterTaxCHF });
  }

  const etfPriceValueEnd = totalEtfUnits * (await etfLook.priceCHFOn(endDate));
  const years = yearsBetween(firstDate, endDate);
  const etfMeanValue = (totalInvestedCHF + etfPriceValueEnd) / 2;
  // Accumulating funds: income reinvested into price but still taxed (docs §3).
  const etfIncomeDrag =
    preTax || !bench.accumulating
      ? 0
      : fundIncomeTaxDrag(etfMeanValue, bench.incomeYield, years, bench.domicile, tax);
  const etfWealthTax = preTax ? 0 : wealthTax(etfMeanValue, years, tax);

  const counterfactualValueCHF =
    etfPriceValueEnd + etfNetDistCHF - etfIncomeDrag - etfWealthTax;

  const etfTerminalFlows: CashFlow[] = [
    ...etfFlowsAfterTax,
    { date: endDate, amount: etfPriceValueEnd - etfIncomeDrag - etfWealthTax },
  ];
  const benchmarkXirr = xirr(etfTerminalFlows);

  // --- Actual holding value at endDate (after tax) ---
  const actual = await actualValueSeries(instrument, sorted, stockLook, tax, preTax, endDate);
  const actualValueCHF = actual.endValueCHF;

  const deltaCHF = actualValueCHF - counterfactualValueCHF;
  const deltaPct = counterfactualValueCHF !== 0 ? deltaCHF / Math.abs(counterfactualValueCHF) : 0;

  // --- Aligned monthly series for the delta chart ---
  const series = await buildSeries(
    firstDate,
    endDate,
    (date) => actual.valueAt(date),
    async (date) => {
      const units = unitsHeldAt(date);
      const price = await etfLook.priceCHFOn(date);
      const distSoFar = await etfDistToDate(etfDivs, etfEvents, bench, tax, preTax, firstDate, date);
      return units * price + distSoFar;
    },
  );

  // Recovery months: if actual ahead, how long for ETF (at its CAGR) to catch actual.
  let recoveryMonths: number | null = null;
  const etfCagr = cagr(totalInvestedCHF, counterfactualValueCHF, years) ?? bench.incomeYield;
  if (deltaCHF > 0 && counterfactualValueCHF > 0 && etfCagr && etfCagr > 0) {
    const monthlyRate = Math.pow(1 + etfCagr, 1 / 12) - 1;
    if (monthlyRate > 0) {
      recoveryMonths = Math.log(actualValueCHF / counterfactualValueCHF) / Math.log(1 + monthlyRate);
    }
  }

  return {
    benchmarkSymbol: bench.symbol,
    benchmarkName: bench.name,
    actualValueCHF,
    counterfactualValueCHF,
    deltaCHF,
    deltaPct,
    actualPreTaxCHF: actual.endValuePreTaxCHF,
    counterfactualPreTaxCHF: etfPriceValueEnd + etfGrossDistCHF,
    actualXirr: actual.xirr,
    benchmarkXirr,
    series,
    recoveryMonths: recoveryMonths && Number.isFinite(recoveryMonths) ? recoveryMonths : null,
  };
}

async function etfDistToDate(
  etfDivs: { date: string; close: number }[],
  etfEvents: { date: string; units: number }[],
  bench: BenchmarkEtf,
  tax: AppSettings['tax'],
  preTax: boolean,
  from: string,
  to: string,
): Promise<number> {
  let sum = 0;
  for (const d of etfDivs) {
    if (d.date < from || d.date > to) continue;
    const units = etfEvents.filter((e) => e.date <= d.date).reduce((s, e) => s + e.units, 0);
    if (units <= 0) continue;
    const fx = await getFxRate(bench.currency, 'CHF', d.date);
    const grossCHF = units * d.close * fx;
    sum += preTax ? grossCHF : dividendTax(grossCHF, bench.domicile, tax).netAfterTaxCHF;
  }
  return sum;
}

interface ActualSeries {
  endValueCHF: number;
  endValuePreTaxCHF: number;
  xirr: number | null;
  valueAt: (date: string) => Promise<number>;
}

async function actualValueSeries(
  instrument: Instrument,
  sorted: Transaction[],
  stockLook: PriceLookup,
  tax: AppSettings['tax'],
  preTax: boolean,
  endDate: string,
): Promise<ActualSeries> {
  // Running units and cumulative after-tax dividends as a function of date.
  const buysSells = sorted.filter((t) => t.action !== 'dividend');
  const divs = sorted.filter((t) => t.action === 'dividend');
  const flows: CashFlow[] = [];

  const unitsAt = (date: string) =>
    buysSells
      .filter((t) => t.date <= date)
      .reduce((s, t) => s + (t.action === 'buy' ? t.quantity : -t.quantity), 0);

  // Cash returned by sells up to `date` (CHF, historical FX). Without this a fully-closed
  // position would look like it vanished to ~0 instead of the cash the sells realised.
  const proceedsToDate = async (date: string) => {
    let sum = 0;
    for (const t of buysSells) {
      if (t.action !== 'sell' || t.date > date) continue;
      const orig = t.quantity * t.unitPrice - (t.fees ?? 0);
      const fx = await getFxRate(t.currency || instrument.currency, 'CHF', t.date);
      sum += orig * fx;
    }
    return sum;
  };

  const divToDate = async (date: string) => {
    let sum = 0;
    for (const d of divs) {
      if (d.date > date) continue;
      const gross = d.grossAmount != null ? d.grossAmount : d.quantity * d.unitPrice;
      const fx = await getFxRate(d.currency || instrument.currency, 'CHF', d.date);
      const grossCHF = gross * fx;
      sum += preTax ? grossCHF : dividendTax(grossCHF, instrument.domicile, tax).netAfterTaxCHF;
    }
    return sum;
  };

  // Cash flows for xirr
  for (const t of buysSells) {
    const orig = t.quantity * t.unitPrice + (t.action === 'buy' ? t.fees ?? 0 : -(t.fees ?? 0));
    const fx = await getFxRate(t.currency || instrument.currency, 'CHF', t.date);
    flows.push({ date: t.date, amount: (t.action === 'buy' ? -1 : 1) * orig * fx });
  }
  for (const d of divs) {
    const gross = d.grossAmount != null ? d.grossAmount : d.quantity * d.unitPrice;
    const fx = await getFxRate(d.currency || instrument.currency, 'CHF', d.date);
    const grossCHF = gross * fx;
    flows.push({
      date: d.date,
      amount: preTax ? grossCHF : dividendTax(grossCHF, instrument.domicile, tax).netAfterTaxCHF,
    });
  }

  const valueAt = async (date: string) => {
    const units = unitsAt(date);
    const priceCHF = units > 0 ? await stockLook.priceCHFOn(date) : 0;
    const div = await divToDate(date);
    const proceeds = await proceedsToDate(date);
    return units * priceCHF + div + proceeds;
  };

  const endUnits = unitsAt(endDate);
  const endPriceCHF = endUnits > 0 ? await stockLook.priceCHFOn(endDate) : 0;
  const endValueCHF = await valueAt(endDate);
  const endValuePreTaxCHF =
    endUnits * endPriceCHF + (await divToDate(endDate)) + (await proceedsToDate(endDate));

  const terminal: CashFlow[] =
    endUnits > 0 ? [...flows, { date: endDate, amount: endUnits * endPriceCHF }] : flows;

  return { endValueCHF, endValuePreTaxCHF, xirr: xirr(terminal), valueAt };
}

async function buildSeries(
  from: string,
  to: string,
  actualAt: (date: string) => Promise<number>,
  benchAt: (date: string) => Promise<number>,
): Promise<CounterfactualPoint[]> {
  const points: CounterfactualPoint[] = [];
  let cursor = dayjs(from).startOf('month');
  const end = dayjs(to);
  const step = end.diff(dayjs(from), 'month') > 120 ? 3 : 1; // coarser for very long histories
  while (cursor.isBefore(end) || cursor.isSame(end, 'month')) {
    const d = cursor.format('YYYY-MM-DD');
    points.push({
      date: d,
      actualCHF: round2(await actualAt(d)),
      benchmarkCHF: round2(await benchAt(d)),
    });
    cursor = cursor.add(step, 'month');
  }
  const last = dayjs(to).format('YYYY-MM-DD');
  points.push({ date: last, actualCHF: round2(await actualAt(last)), benchmarkCHF: round2(await benchAt(last)) });
  return points;
}

function round2(n: number) {
  return Math.round(n * 100) / 100;
}

function emptyResult(bench: BenchmarkEtf): CounterfactualResult {
  return {
    benchmarkSymbol: bench.symbol,
    benchmarkName: bench.name,
    actualValueCHF: 0,
    counterfactualValueCHF: 0,
    deltaCHF: 0,
    deltaPct: 0,
    actualPreTaxCHF: 0,
    counterfactualPreTaxCHF: 0,
    actualXirr: null,
    benchmarkXirr: null,
    series: [],
    recoveryMonths: null,
  };
}

File: backend/app/services/counterfactual.py
"""Counterfactual engine — port of counterfactual.ts.

Mirrors every cash outflow into the stock as a purchase of the benchmark ETF on the same
date, rolls it forward, and returns the CHF opportunity-cost delta, after Swiss tax."""
from __future__ import annotations

import math

import pandas as pd

from ..reference.geo import country_from_symbol
from .finance_math import cagr, xirr, years_between
from .fx import ensure_fx_range, get_fx_rate
from .marketdata import ensure_history, get_dividends, get_quote, price_on
from .tax import dividend_tax, fund_income_tax_drag, wealth_tax


def _resolve_benchmark(symbol: str, settings: dict) -> dict:
    for b in settings["benchmarks"]:
        if b["symbol"] == symbol:
            return b
    return {
        "symbol": symbol, "name": symbol, "currency": "USD",
        "domicile": country_from_symbol(symbol) or "US",
        "incomeYield": settings["tax"]["defaultEtfIncomeYield"], "accumulating": False,
    }


class _Lookup:
    def __init__(self, ccy: str, symbol: str):
        self.ccy = ccy
        self._symbol = symbol

    def price_chf_on(self, date: str) -> float:
        p = price_on(self._symbol, date)
        if p is None:
            return 0.0
        return p * get_fx_rate(self.ccy, "CHF", date)


def _make_lookup(symbol: str, from_date: str, fallback_ccy: str) -> _Lookup:
    ensure_history(symbol, from_date)
    try:
        q = get_quote(symbol)
        ccy = q["currency"] or fallback_ccy
    except Exception:  # noqa: BLE001
        ccy = fallback_ccy
    return _Lookup(ccy, symbol)


def _round2(n: float) -> float:
    return round(n * 100) / 100


def _empty_result(bench: dict) -> dict:
    return {
        "benchmarkSymbol": bench["symbol"], "benchmarkName": bench["name"],
        "actualValueCHF": 0, "counterfactualValueCHF": 0, "deltaCHF": 0, "deltaPct": 0,
        "actualPreTaxCHF": 0, "counterfactualPreTaxCHF": 0,
        "actualXirr": None, "benchmarkXirr": None, "series": [], "recoveryMonths": None,
    }


def compute_counterfactual(instrument: dict, txs: list[dict], benchmark_symbol: str,
                          settings: dict, pre_tax: bool = False, as_of: str | None = None) -> dict:
    tax = settings["tax"]
    bench = _resolve_benchmark(benchmark_symbol, settings)
    today = as_of or pd.Timestamp.utcnow().strftime("%Y-%m-%d")
    sorted_txs = sorted(
        [t for t in txs if t.get("category") != "corporate_action" and t["date"] <= today],
        key=lambda t: t["date"],
    )
    buys = [t for t in sorted_txs if t["action"] == "buy"]
    sells = [t for t in sorted_txs if t["action"] == "sell"]

    if not buys:
        return _empty_result(bench)

    first_date = sorted_txs[0]["date"]
    bought_qty = sum(t["quantity"] for t in buys)
    sold_qty = sum(t["quantity"] for t in sells)
    fully_closed = sold_qty >= bought_qty - 1e-9 and len(sells) > 0
    end_date = sells[-1]["date"] if fully_closed else today
    if pd.Timestamp(end_date) > pd.Timestamp(today):
        end_date = today

    currencies = {"CHF", instrument["currency"], bench["currency"]}
    for t in txs:
        currencies.add(t.get("currency"))
    ensure_fx_range([c for c in currencies if c], first_date, end_date)

    stock_look = _make_lookup(instrument["symbol"], first_date, instrument["currency"])
    etf_look = _make_lookup(bench["symbol"], first_date, bench["currency"])

    # --- Build ETF counterfactual units by mirroring each buy's CHF outflow ---
    etf_events: list[dict] = []
    etf_flows_after_tax: list[dict] = []
    for b in buys:
        cost_orig = b["quantity"] * b["unitPrice"] + (b.get("fees") or 0)
        fx_buy = get_fx_rate(b.get("currency") or instrument["currency"], "CHF", b["date"])
        cash_chf = cost_orig * fx_buy
        etf_price_chf = etf_look.price_chf_on(b["date"])
        units = cash_chf / etf_price_chf if etf_price_chf > 0 else 0
        etf_events.append({"date": b["date"], "units": units, "cashCHF": cash_chf})
        etf_flows_after_tax.append({"date": b["date"], "amount": -cash_chf})
    total_invested_chf = sum(e["cashCHF"] for e in etf_events)

    def units_held_at(date: str) -> float:
        return sum(e["units"] for e in etf_events if e["date"] <= date)

    total_etf_units = units_held_at(end_date)

    etf_divs = get_dividends(bench["symbol"], first_date)
    etf_net_dist_chf = 0.0
    etf_gross_dist_chf = 0.0
    for d in etf_divs:
        if d["date"] < first_date or d["date"] > end_date:
            continue
        units = units_held_at(d["date"])
        if units <= 0:
            continue
        fx = get_fx_rate(bench["currency"], "CHF", d["date"])
        gross_chf = units * d["close"] * fx
        etf_gross_dist_chf += gross_chf
        bd = dividend_tax(gross_chf, bench["domicile"], tax)
        etf_net_dist_chf += gross_chf if pre_tax else bd.netAfterTaxCHF
        etf_flows_after_tax.append({"date": d["date"],
                                    "amount": gross_chf if pre_tax else bd.netAfterTaxCHF})

    etf_price_value_end = total_etf_units * etf_look.price_chf_on(end_date)
    years = years_between(first_date, end_date)
    etf_mean_value = (total_invested_chf + etf_price_value_end) / 2
    etf_income_drag = (
        0 if (pre_tax or not bench["accumulating"])
        else fund_income_tax_drag(etf_mean_value, bench["incomeYield"], years, bench["domicile"], tax)
    )
    etf_wealth_tax = 0 if pre_tax else wealth_tax(etf_mean_value, years, tax)

    counterfactual_value_chf = etf_price_value_end + etf_net_dist_chf - etf_income_drag - etf_wealth_tax

    etf_terminal_flows = [
        *etf_flows_after_tax,
        {"date": end_date, "amount": etf_price_value_end - etf_income_drag - etf_wealth_tax},
    ]
    benchmark_xirr = xirr(etf_terminal_flows)

    # --- Actual holding value at endDate (after tax) ---
    actual = _actual_value_series(instrument, sorted_txs, stock_look, tax, pre_tax, end_date)
    actual_value_chf = actual["endValueCHF"]

    delta_chf = actual_value_chf - counterfactual_value_chf
    delta_pct = delta_chf / abs(counterfactual_value_chf) if counterfactual_value_chf != 0 else 0

    def bench_at(date: str) -> float:
        units = units_held_at(date)
        price = etf_look.price_chf_on(date)
        dist_so_far = _etf_dist_to_date(etf_divs, etf_events, bench, tax, pre_tax, first_date, date)
        return units * price + dist_so_far

    series = _build_series(first_date, end_date, actual["valueAt"], bench_at)

    recovery_months = None
    etf_cagr = cagr(total_invested_chf, counterfactual_value_chf, years)
    if etf_cagr is None:
        etf_cagr = bench["incomeYield"]
    if delta_chf > 0 and counterfactual_value_chf > 0 and etf_cagr and etf_cagr > 0:
        monthly_rate = (1 + etf_cagr) ** (1 / 12) - 1
        if monthly_rate > 0:
            recovery_months = math.log(actual_value_chf / counterfactual_value_chf) / math.log(1 + monthly_rate)

    return {
        "benchmarkSymbol": bench["symbol"],
        "benchmarkName": bench["name"],
        "actualValueCHF": actual_value_chf,
        "counterfactualValueCHF": counterfactual_value_chf,
        "deltaCHF": delta_chf,
        "deltaPct": delta_pct,
        "actualPreTaxCHF": actual["endValuePreTaxCHF"],
        "counterfactualPreTaxCHF": etf_price_value_end + etf_gross_dist_chf,
        "actualXirr": actual["xirr"],
        "benchmarkXirr": benchmark_xirr,
        "series": series,
        "recoveryMonths": recovery_months if (recovery_months is not None and math.isfinite(recovery_months)) else None,
    }


def _etf_dist_to_date(etf_divs, etf_events, bench, tax, pre_tax, from_date, to) -> float:
    total = 0.0
    for d in etf_divs:
        if d["date"] < from_date or d["date"] > to:
            continue
        units = sum(e["units"] for e in etf_events if e["date"] <= d["date"])
        if units <= 0:
            continue
        fx = get_fx_rate(bench["currency"], "CHF", d["date"])
        gross_chf = units * d["close"] * fx
        total += gross_chf if pre_tax else dividend_tax(gross_chf, bench["domicile"], tax).netAfterTaxCHF
    return total


def _actual_value_series(instrument, sorted_txs, stock_look, tax, pre_tax, end_date) -> dict:
    buys_sells = [t for t in sorted_txs if t["action"] != "dividend"]
    divs = [t for t in sorted_txs if t["action"] == "dividend"]
    flows: list[dict] = []

    def units_at(date: str) -> float:
        return sum(
            (t["quantity"] if t["action"] == "buy" else -t["quantity"])
            for t in buys_sells if t["date"] <= date
        )

    def proceeds_to_date(date: str) -> float:
        total = 0.0
        for t in buys_sells:
            if t["action"] != "sell" or t["date"] > date:
                continue
            orig = t["quantity"] * t["unitPrice"] - (t.get("fees") or 0)
            fx = get_fx_rate(t.get("currency") or instrument["currency"], "CHF", t["date"])
            total += orig * fx
        return total

    def div_to_date(date: str) -> float:
        total = 0.0
        for d in divs:
            if d["date"] > date:
                continue
            gross = d["grossAmount"] if d.get("grossAmount") is not None else d["quantity"] * d["unitPrice"]
            fx = get_fx_rate(d.get("currency") or instrument["currency"], "CHF", d["date"])
            gross_chf = gross * fx
            total += gross_chf if pre_tax else dividend_tax(gross_chf, instrument.get("domicile"), tax).netAfterTaxCHF
        return total

    for t in buys_sells:
        orig = t["quantity"] * t["unitPrice"] + ((t.get("fees") or 0) if t["action"] == "buy" else -(t.get("fees") or 0))
        fx = get_fx_rate(t.get("currency") or instrument["currency"], "CHF", t["date"])
        flows.append({"date": t["date"], "amount": (-1 if t["action"] == "buy" else 1) * orig * fx})
    for d in divs:
        gross = d["grossAmount"] if d.get("grossAmount") is not None else d["quantity"] * d["unitPrice"]
        fx = get_fx_rate(d.get("currency") or instrument["currency"], "CHF", d["date"])
        gross_chf = gross * fx
        flows.append({"date": d["date"],
                      "amount": gross_chf if pre_tax else dividend_tax(gross_chf, instrument.get("domicile"), tax).netAfterTaxCHF})

    def value_at(date: str) -> float:
        units = units_at(date)
        price_chf = stock_look.price_chf_on(date) if units > 0 else 0
        return units * price_chf + div_to_date(date) + proceeds_to_date(date)

    end_units = units_at(end_date)
    end_price_chf = stock_look.price_chf_on(end_date) if end_units > 0 else 0
    end_value_chf = value_at(end_date)
    end_value_pre_tax_chf = end_units * end_price_chf + div_to_date(end_date) + proceeds_to_date(end_date)

    terminal = [*flows, {"date": end_date, "amount": end_units * end_price_chf}] if end_units > 0 else flows

    return {
        "endValueCHF": end_value_chf,
        "endValuePreTaxCHF": end_value_pre_tax_chf,
        "xirr": xirr(terminal),
        "valueAt": value_at,
    }


def _build_series(from_date, to, actual_at, bench_at) -> list[dict]:
    points: list[dict] = []
    cursor = pd.Timestamp(from_date).to_period("M").to_timestamp()  # startOf month
    end = pd.Timestamp(to)
    step = 3 if (end.to_period("M") - pd.Timestamp(from_date).to_period("M")).n > 120 else 1
    while cursor < end or cursor.to_period("M") == end.to_period("M"):
        d = cursor.strftime("%Y-%m-%d")
        points.append({"date": d, "actualCHF": _round2(actual_at(d)), "benchmarkCHF": _round2(bench_at(d))})
        cursor = cursor + pd.DateOffset(months=step)
    last = pd.Timestamp(to).strftime("%Y-%m-%d")
    points.append({"date": last, "actualCHF": _round2(actual_at(last)), "benchmarkCHF": _round2(bench_at(last))})
    return points
